tools/api.py

```diff
@@ -0,0 +1,837 @@
+        self.name = name
+        self.ctype = ctype
+        self.cpptype = cpptype
+        self.cpp_class = CPPClass(name, ctype)
+        add_handle(name, ctype, cpptype)
+        cpp_type_map[cpptype] = name
+
+    def cname(self, name):
+        return self.ctype + '_' + name
+
+    def substitute(self, s, **kwargs):
+        return Template(s).safe_substitute(name=self.name,
+                                           ctype=self.ctype,
+                                           cpptype=self.cpptype,
+                                           **kwargs)
+
+    def constructor(self, name, params=None, fname=None, invoke=None,
+                    **kwargs):
+        create = self.substitute('allocate<${cpptype}>($@)')
+        if fname:
+            create = self.substitute('allocate<${cpptype}>(${fname}($@))',
+                                     fname=fname)
+
+        f = add_function(self.cname(name),
+                         params=params,
+                         invoke=invoke or create,
+                         returns=self.cpptype + '*',
+                         return_name=self.name,
+                         **kwargs)
+        self.cpp_class.add_constructor(name, f)
+        return self
+
+    def method(self,
+               name,
+               params=None,
+               fname=None,
+               invoke=None,
+               cpp_name=None,
+               const=None,
+               **kwargs):
+        cpptype = self.cpptype
+        if const:
+            cpptype = Type(cpptype).add_const().str()
+        p = Parameter(self.name, cpptype)
+        args = to_template_vars(params or [])
+        f = add_function(self.cname(name),
+                         params=[p] + (params or []),
+                         invoke=invoke
+                         or self.substitute('${var}.${fname}(${args})',
+                                            var=template_var(self.name),
+                                            fname=fname or name,
+                                            args=args),
+                         **kwargs)
+        self.cpp_class.add_method(cpp_name or name, f)
+        return self
+
+    def function(self, name, params=None, **kwargs):
+        add_function(self.cname(name), params=params, **kwargs)
+        return self
+
+    def add_cpp_class(self):
+        cpp_classes.append(self.cpp_class)
+
+
+def handle(ctype, cpptype, name=None):
+    def with_handle(f):
+        n = name or f.__name__
+        h = Handle(n, ctype, cpptype)
+        f(h)
+        h.add_cpp_class()
+
+        @wraps(f)
+        def decorated(*args, **kwargs):
+            return f(*args, **kwargs)
+
+        return decorated
+
+    return with_handle
+
+
+def template_eval(template, **kwargs):
+    start = '<%'
+    end = '%>'
+    escaped = (re.escape(start), re.escape(end))
+    mark = re.compile('%s(.*?)%s' % escaped, re.DOTALL)
+    for key in kwargs:
+        exec ('%s = %s' % (key, kwargs[key]))
```

Comment:
you can fix the format error by making sure "yapf" is pointing to the python3 version when doing formatting locally